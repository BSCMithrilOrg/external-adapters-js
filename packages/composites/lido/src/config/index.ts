import { Requester, util } from '@chainlink/ea-bootstrap'
import { Config as DefaultConfig } from '@chainlink/types'

export const DEFAULT_ENDPOINT = 'exchangerate'
export const NAME = 'LIDO'
export const DEFAULT_WSTETH_ADDRESS = '0x7f39c581f595b53c5cb19bd0b3f8da6c935e2ca0'
export const FLOATING_POINT_DECIMALS = 18

export interface Config extends DefaultConfig {
  wstEthAddress: string
}

export const makeConfig = (prefix?: string): Config => {
  const wstEthAddress = util.getEnv('WSTETH_ADDRESS', prefix) || DEFAULT_WSTETH_ADDRESS
  return {
    ...Requester.getDefaultConfig(),
    wstEthAddress,
    defaultEndpoint: DEFAULT_ENDPOINT,
  }
}
