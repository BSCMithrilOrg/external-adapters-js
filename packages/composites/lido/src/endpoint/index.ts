export * as exchangerate from './exchangerate'
