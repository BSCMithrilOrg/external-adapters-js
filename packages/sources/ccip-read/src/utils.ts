import { Fragment, Interface, JsonFragment } from '@ethersproject/abi'

export const toInterface = (
  abi: string | readonly (string | Fragment | JsonFragment)[] | Interface,
): Interface => {
  return Interface.isInterface(abi) ? abi : new Interface(abi)
}
