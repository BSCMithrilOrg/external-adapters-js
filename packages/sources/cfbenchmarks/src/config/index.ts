import { Requester, util } from '@chainlink/ea-bootstrap'
import { Config as BaseConfig } from '@chainlink/types'

export const NAME = 'CFBENCHMARKS'

export const ENV_API_USERNAME = 'API_USERNAME'
export const ENV_API_PASSWORD = 'API_PASSWORD'
export const ENV_API_SECONDARY = 'API_SECONDARY'

export const DEFAULT_ENDPOINT = 'values'
export const DEFAULT_API_ENDPOINT = 'https://www.cfbenchmarks.com/api'
export const DEFAULT_WS_API_ENDPOINT = 'wss://www.cfbenchmarks.com/ws/v4'
export const SECONDARY_API_ENDPOINT = 'https://unregprod.cfbenchmarks.com/api'
export const SECONDARY_WS_API_ENDPOINT = 'wss://unregprod.cfbenchmarks.com/ws/v4'

export type Config = BaseConfig & {
  useSecondary: boolean
  username: string
  password: string
}

export const makeConfig = (prefix?: string): Config => {
  const config = {
    ...Requester.getDefaultConfig(prefix),
    useSecondary: util.parseBool(util.getEnv(ENV_API_SECONDARY)),
    username: util.getRequiredEnv(ENV_API_USERNAME, prefix),
    password: util.getRequiredEnv(ENV_API_PASSWORD, prefix),
  }

  config.api.baseURL =
    config.api.baseURL || (config.useSecondary ? SECONDARY_API_ENDPOINT : DEFAULT_API_ENDPOINT)
  config.api.baseWsURL =
    config.api.baseWsURL ||
    (config.useSecondary ? SECONDARY_WS_API_ENDPOINT : DEFAULT_WS_API_ENDPOINT)

  config.defaultEndpoint = DEFAULT_ENDPOINT
  return config
}
