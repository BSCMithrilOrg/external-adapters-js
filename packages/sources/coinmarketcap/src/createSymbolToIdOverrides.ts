import * as process from 'process'
import * as fs from 'fs'

import axios from 'axios'

type OverrideObj = {
  [adapterName: string]: {
    [symbol: string]: string
  }
}

type CoinData = {
  id: number
  name: string
  symbol: string
  slug: string
  rank: number
  [otherData: string]: unknown
}

export const getSymbolToIdOverrides = async (): Promise<OverrideObj> => {
  if (!process.env.API_KEY) throw Error('API_KEY environment variable not set.')
  const headers = { 'X-CMC_PRO_API_KEY': process.env.API_KEY }
  const response = await axios.get('https://pro-api.coinmarketcap.com/v1/cryptocurrency/map', {
    headers,
  })
  const coinsList = response.data.data
  const adapterSymToIdOverrides = {} as { [symbol: string]: string }

  coinsList.forEach((coinData: CoinData) => {
    adapterSymToIdOverrides[coinData.symbol] = coinData.id.toString()
  })

  return { coinmarketcap: adapterSymToIdOverrides }
}

const createSymbolToIdOverridesFile = async (): Promise<void> => {
  const symbolToIdOverrides = await getSymbolToIdOverrides()
  fs.writeFileSync('symbolToIdOverrides.json', JSON.stringify(symbolToIdOverrides), {
    encoding: 'utf-8',
  })
}

createSymbolToIdOverridesFile()
