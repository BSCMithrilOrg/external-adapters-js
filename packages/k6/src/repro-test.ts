import { check } from 'k6'
import { SharedArray } from 'k6/data'
import http, { BatchRequests } from 'k6/http'
import { Rate } from 'k6/metrics'
import { Payload } from './config/types'

// load the test iterations from the environment or default to 100
let iterations = 20000
if (__ENV.LOAD_TEST_ITERATIONS) {
  iterations = parseInt(__ENV.LOAD_TEST_ITERATIONS)
}

// set the k6 running options
export const options = {
  vus: 4,
  iterations,
  thresholds: {
    http_req_failed: ['rate<0.01'], // http errors should be less than 1%
    http_req_duration: ['p(95)<200'], // 95% of requests should be below 200ms
  },
}

export const errorRate = new Rate('errors')

// load the test data, if data was generated then load it from the generated file
if (!__ENV.PAYLOAD_PATH) {
  console.error('PAYLOAD_PATH env var is not present')
  process.exit(1)
}

const payloadData: Payload[] = new SharedArray('payloadData', () =>
  JSON.parse(open(__ENV.PAYLOAD_PATH)),
)

function buildRequests() {
  const batchRequests: BatchRequests = {}
  const params = {
    headers: {
      'Content-Type': 'application/json',
    },
  }

  const adapterName = __ENV.LOAD_TEST_ADAPTER_NAME
  const adapterUrl = `http://host.docker.internal:8080`

  for (const payload of payloadData) {
    batchRequests[`${adapterName}-${payload.name}`] = {
      method: payload.method,
      url: adapterUrl,
      body: JSON.stringify({
        data: JSON.parse(payload.data),
      }),
      params,
    }
  }

  return batchRequests
}

const batchRequests = buildRequests()

export default (): void => {
  const responses = http.batch(batchRequests)
  for (const [name, response] of Object.entries(responses)) {
    const result = check(response, {
      [`${name} returned 200`]: (r) => r.status == 200,
    })

    errorRate.add(!result)
  }
}
