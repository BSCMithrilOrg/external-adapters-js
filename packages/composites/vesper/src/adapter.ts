import { AdapterData, APIEndpoint, Builder } from '@chainlink/ea-bootstrap'
import * as TA from '@chainlink/token-allocation-adapter'
import type { ExecuteWithConfig, ExecuteFactory } from '@chainlink/ea-bootstrap'
import { makeConfig, Config } from './config'
import * as endpoints from './endpoint'

export const execute: ExecuteWithConfig<Config, endpoints.TInputParameters> = async (
  request,
  context,
  config,
) => {
  return Builder.buildSelector<Config, endpoints.TInputParameters>(
    request,
    context,
    config,
    endpoints,
  )
}

export const endpointSelector = TA.makeEndpointSelector(
  makeConfig,
  endpoints as Record<string, APIEndpoint<Config, AdapterData>>,
  'price',
)
// TODO: composite endpoint types

export const makeExecute: ExecuteFactory<Config, endpoints.TInputParameters> = (config) => {
  return async (request, context) => execute(request, context, config || makeConfig())
}
