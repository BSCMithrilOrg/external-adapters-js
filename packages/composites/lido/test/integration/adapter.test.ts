import request, { SuperTest, Test } from 'supertest'
import * as process from 'process'
import { server as startServer } from '../../src/index'
import nock from 'nock'
import * as http from 'http'
import { AddressInfo } from 'net'
import { mockContractCallResponseSuccess } from './fixtures'

describe('LIDO', () => {
  process.env.RPC_URL = 'http://localhost:8545'
  mockContractCallResponseSuccess()

  let server: http.Server
  let req: SuperTest<Test>

  beforeAll(async () => {
    process.env.CACHE_ENABLED = 'false'
    if (process.env.RECORD) {
      nock.recorder.rec()
    }
    server = await startServer()
    req = request(`localhost:${(server.address() as AddressInfo).port}`)
  })

  it('fetches the amount of stETH for 1 wstETH', async () => {
    const response = await req
      .post('/')
      .send({})
      .set('Accept', '*/*')
      .set('Content-Type', 'application/json')
      .expect('Content-Type', /json/)
      .expect(200)
    expect(response.body).toMatchSnapshot()
  })

  afterAll((done) => {
    if (process.env.RECORD) {
      nock.recorder.play()
    }

    server.close(done)
    nock.restore()
    nock.cleanAll()
    nock.enableNetConnect()
  })
})
