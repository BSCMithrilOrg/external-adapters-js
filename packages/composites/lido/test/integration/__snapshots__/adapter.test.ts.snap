// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LIDO fetches the amount of stETH for 1 wstETH 1`] = `
Object {
  "data": Object {
    "result": 1065508199231399700,
  },
  "jobRunID": "1",
  "providerStatusCode": 200,
  "result": 1065508199231399700,
  "statusCode": 200,
}
`;
